datafusion-federation/src/analyzer/mod.rs

```diff
@@ -18,21 +18,20 @@ use datafusion::{
 use scan_result::ScanResult;
 use std::collections::HashMap;
 use std::sync::Arc;
-use std::sync::RwLock;
 
+pub use optimize_projections::OptimizeProjections;
+
+mod optimize_projections;
+
+/// An analyzer rule to identifying sub-plans to federate
+///
+/// The analyzer logic walks over the plan, look for the largest subtrees that only have
+/// TableScans from the same [`FederationProvider`]. The 'largest sub-trees' are passed to their
+/// respective [`FederationProvider::analyzer`].
 #[derive(Debug)]
 pub struct FederationAnalyzerRule {
+    // Optimization rules to run before the federated plan is created
     optimizer: Optimizer,
-    provider_map: Arc<RwLock<HashMap<TableReference, ScanResult>>>,
```

Comment:
This was a huge bug in the current implementation. The Analyzer rule was keeping state across runs of tablereferences from previous queries!!